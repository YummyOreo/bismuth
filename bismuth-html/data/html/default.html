<!DOCTYPE html>
<html lang="">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <meta name="head:count" content="4">
    <link rel="stylesheet" href="/assets/css/colors.css">
    <title>{title}</title>

    <style>
        * {
            color: var(--text-1);
        }
        body {
            background-color: var(--background-1);
        }
    </style>
</head>

<body>
    {elements}
</body>

</html>
