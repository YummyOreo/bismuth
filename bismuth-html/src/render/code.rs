use syntect::easy::HighlightLines;
use syntect::highlighting::{Color, ThemeSet};
use syntect::html::highlighted_html_for_string;
use syntect::parsing::SyntaxSet;

use syntect::Error;
use thiserror::Error;

fn init() -> (SyntaxSet, ThemeSet) {
    let ps = SyntaxSet::load_defaults_newlines();
    let ts = ThemeSet::load_defaults();

    (ps, ts)
}

// #[derive(Error, Debug)]
// #[error("Could not find lang: `{lang}`")]
// pub struct FindLangError {
//     lang: String,
// }

#[derive(Debug, Error)]
pub enum HighlightError {
    #[error("Could not find lang: `{lang}`")]
    FindLangError { lang: String },
    #[error("Internal error {0}")]
    Internal(#[from] Error),
}

pub fn highlight(lang: String, code: String) -> Result<String, HighlightError> {
    let (ps, ts) = init();
    let syntax = match ps.find_syntax_by_extension(&lang) {
        Some(syntax) => syntax,
        None => ps
            .find_syntax_by_name(&lang)
            .ok_or(HighlightError::FindLangError { lang })?,
    };

    highlighted_html_for_string(&code, &ps, syntax, &ts.themes["InspiredGitHub"])
        .map_err(HighlightError::Internal)
}

#[cfg(test)]
mod test {

    use super::*;
    macro_rules! snapshot {
        ($content:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../../testdata/output/render/code/");
            settings.bind(|| {
                insta::assert_snapshot!($content);
            });
        };
    }

    #[test]
    fn test() {
        let code = "fn test() {\n\tlet a = 10;\n}".to_string();
        let html = highlight("rs".to_string(), code).unwrap();
        snapshot!(html);
    }

    #[test]
    fn test_inline() {
        let code = "let test = 10".to_string();
        let html = highlight("rs".to_string(), code).unwrap();
        snapshot!(html);
    }
}
