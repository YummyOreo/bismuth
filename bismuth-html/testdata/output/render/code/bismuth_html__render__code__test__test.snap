---
source: bismuth-html/src/render/code.rs
expression: html
---
<pre style="background-color:#3b3228;">
<span style="color:#a89bb9;">fn </span><span style="color:#8ab3b5;">test</span><span style="color:#d0c8c6;">() {
</span><span style="color:#d0c8c6;">	</span><span style="color:#a89bb9;">let</span><span style="color:#d0c8c6;"> a = </span><span style="color:#d28b71;">10</span><span style="color:#d0c8c6;">;
</span><span style="color:#d0c8c6;">}</span></pre>

